### PR TITLE
gradle: 8.4 -> 8.5

### PR DESCRIPTION
**Pull request by liff**
## Description of changes

[New Gradle release](https://docs.gradle.org/8.5/release-notes.html).

Now that it officially supports Java 21, also bump `defaultJava`.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### Priorities

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/development/tools/build-managers/gradle/default.nix [hunk 1]
```diff
@@ -128,10 +128,10 @@ rec {
   # https://docs.gradle.org/current/userguide/compatibility.html
 
   gradle_8 = gen {
-    version = "8.4";
+    version = "8.5";
     nativeVersion = "0.22-milestone-25";
-    sha256 = "1bkjxw7i0lm17pdyyvka4xpl6z0cdj0izagphync6839i2pg66iy";
-    defaultJava = jdk17;
+    sha256 = "09hhlka9vc4xr8xa68d8qcvyhsaaicrmh210x0wif23a0s3ng4lx";
```
Comment (kirillrdy) [inline pkgs/development/tools/build-managers/gradle/default.nix:133]:
```suggestion
    hash = "09hhlka9vc4xr8xa68d8qcvyhsaaicrmh210x0wif23a0s3ng4lx";
```

you will also need to change hash to SRI representation, and update param to `gen` function and inherit inside `gen`

Comment (liff) [reply pkgs/development/tools/build-managers/gradle/default.nix]: Changed. The update script _should_ also be changed, but it uses `nix-prefetch-url`, which doesn’t output SRI format (yet?).

Comment (kirillrdy) [reply pkgs/development/tools/build-managers/gradle/default.nix]: if this is too many changes, feel free to do it part of different PR, also If you wish I am happy to make that PR

**Review by lorenzleutgeb — APPROVED** [review]

**Comment by liff** [comment]
Mindustry fails to build on Java 21. Should I revert Gradle’s default Java version or add `gradle = gradle.override { java = openjdk17; };` to mindustry?

**Comment by lorenzleutgeb** [comment]
Mindustry takes the argument `jdk`: https://github.com/NixOS/nixpkgs/blob/4552f3aa254e5c50406ca59c2c04483ef8539161/pkgs/games/mindustry/default.nix#L7

And seems to use it to run Mindustry: https://github.com/NixOS/nixpkgs/blob/4552f3aa254e5c50406ca59c2c04483ef8539161/pkgs/games/mindustry/default.nix#L198

So in the outermost `let` you could introduce `gradle = gradle.override { java = jdk; };` so that Gradle uses the same JDK as the rest of the derivation.

**Comment by liff** [comment]
@lorenzleutgeb that makes sense, thanks.

**Review by fgaz — COMMENTED** [review]
The mindustry change looks good (if it works. I didn't test it)

**Comment by viluon** [comment]
Are there any blockers to merging this?

**Review by lorenzleutgeb — APPROVED** [review]

**Review by viluon — APPROVED** [review]

**Review by kirillrdy — APPROVED** [review]
Result of `nixpkgs-review pr 271141` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>27 packages built:</summary>
  <ul>
    <li>aeron</li>
    <li>apkleaks</li>
    <li>apkleaks.dist</li>
    <li>diffoscope</li>
    <li>diffoscope.dist</li>
    <li>diffoscope.man</li>
    <li>experienced-pixel-dungeon</li>
    <li>gradle (gradle_8)</li>
    <li>gscan2pdf</li>
    <li>gscan2pdf.man</li>
    <li>jabref</li>
    <li>jadx</li>
    <li>jextract</li>
    <li>kotlin-language-server</li>
    <li>mindustry</li>
    <li>mindustry-server</li>
    <li>mindustry-wayland</li>
    <li>pdf-sign</li>
    <li>pdfchain</li>
    <li>pdftk</li>
    <li>pidginPackages.purple-signald</li>
    <li>rat-king-adventure</li>
    <li>rkpd2</li>
    <li>shattered-pixel-dungeon</li>
    <li>shorter-pixel-dungeon</li>
    <li>signald</li>
    <li>signaturepdf</li>
  </ul>
</details>

**Comment by fgaz** [comment]
Seems like there are [breaking changes](https://docs.gradle.org/8.5/userguide/upgrading_version_8.html#changes_8.5) even though it's a minor version, so no backport I guess?

**Comment by liff** [comment]
There have been potential breaking changes in pretty much every single minor Gradle release, so I’m not sure how seriously breaking those actually are 😅 

However, if a backport were to be made, we might want to keep the JDK at 17 in it?

**Comment by nifoc** [comment]
I think this PR broke `signald` for me. It builds fine, but running it crashes with:

```
Dec 14 18:00:51 tanker systemd[1]: Started A daemon for interacting with the Signal Private Messenger.
Dec 14 18:00:51 tanker signald[3635302]: Error: LinkageError occurred while loading main class io.finn.signald.Main
Dec 14 18:00:51 tanker signald[3635302]:         java.lang.UnsupportedClassVersionError: io/finn/signald/Main has been compiled by a more recent version of the Java Runtime (class file version 65.0), this version of the Java Runtime only recognizes class file versions up to 61.0
```

**Comment by liff** [comment]
Looks like signald needs the same treatment as mindustry.

**Comment by liff** [comment]
@nifoc fix in #274279.

**Comment by kirillrdy** [comment]
@nifoc for signald, would something like version test catch the issue ?

**Comment by nifoc** [comment]
> @nifoc for signald, would something like version test catch the issue ?

Looks like it!

```
[root@tanker:~]# nix-shell -p signald

[nix-shell:~]# signald --version
Error: LinkageError occurred while loading main class io.finn.signald.Main
        java.lang.UnsupportedClassVersionError: io/finn/signald/Main has been compiled by a more recent version of the Java Runtime (class file version 65.0), this version of the Java Runtime only recognizes class file versions up to 61.0
```

**Pull request merged**